### ISSUE TITLE
Channel.isWritable() not taking isActive() into account?

### ISSUE THREAD

**Issue by daschl** [netty/netty#3300]
Hi folks,

while debugging an issue I found something that is probably worth discussing.

So the documentation for isWritable() says:

``` java
    /**
     * Returns {@code true} if and only if the I/O thread will perform the
     * requested write operation immediately.  Any write requests made when
     * this method returns {@code false} are queued until the I/O thread is
     * ready to process the queued write requests.
     */
    boolean isWritable();
```

Now I checked the implementation and if I'm not mistaken, this does only check on the ByteBuf but not on the channel status? So could it happen that isWritable returns true, but isActive is actually false?

If so, that would violate the documentation contract, since the op cannot be performed immediately.

So if the user would actually have to check `if(channel.isWritable() && channel.isActive())` before writing, just checking for `channel.isWritable()` is not enough?

I wonder if `isWritable()` should automatically check `isActive()` as well.

**Comment by trustin** [comment]
We could initialize `outboundBuffer` when channel becomes active, so that an inactive channel's `outboundBuffer` is always `null`.

**Comment by normanmaurer** [comment]
Sounds good to me

**Comment by daschl** [comment]
Would this have the same semantics as isActive() over time? I guess my fear is that the assumption can become invalid over time?

**Comment by trustin** [comment]
Yeah, `outboundBuffer != null` could have the same semantics with `isActive()` if we are careful enough. Putting the implementation-level mistakes aside, I think they should be identical, because by definition 'active' means 'being able to send or receive something'. We might have a read-only channel implementation in the future, but we can easily override the default behavior in such a case.

**Comment by normanmaurer** [comment]
I think the ship has sailed and we should just keep it as it is
